app/controllers/application_controller.rb

```diff
@@ -2387,4 +2389,59 @@ def item_sequence_base(asset_type, asset_id)
       return result
     end
   end
+
+  private
+  def extra_logging
+    # Clean out things we can't JSONify so it doesn't crash.
+    def filter_hash(main_hash)
+      main_hash.each_with_object({}) do |(k, v), h|
+        # remove cookie values so we don't clutter the logs
+        if(k.to_s.downcase.include? 'cookie')
+          next
+        end
+        if(v.is_a?(Hash))
+          ch = filter_hash(v)
+          h[k] = ch
+        elsif(v.is_a?(Array))
+          h[k] = v.map do |e|
+            e.is_a?(Hash) ? filter_hash(e) : e
+          end
+        elsif(v.is_a?(String))
+          h[k] = v
+        elsif(v.is_a?(Integer))
+          h[k] = v
+        end
+      end
+    end
+
+    data = {
+      'request' => {
+        'method' => request.method,
+        'headers' => filter_hash(request.headers.env),
+        'ip' => request.ip,
+        'fullpath' => request.fullpath,
+        'remote_ip' => request.remote_ip,
+        'id' => RequestContextGenerator.request_id,
+      },
+      'api_request' => @api_request,
+      'action_name' => @_action_name,
+      'session' => session.to_hash,
+      'response' => {
+        'code' => response.code,
+      },
+      'user' => @current_user,
+    }
```

Comment:
What if this throws? There is enough logic in here that I'm not confident it will never throw and if it does I think we're going to be in bad shape, so I think it's worth wrapping this in a `try/rescue` and set `data` to something that will help diagnose that if we run into it.

Otherwise, this is really cool and I think it'll help us a TON going forward!

Reply:
good note. If I understand correctly, the rescue block below will catch things that happen here, but then it will probably break if data isn't set. I'll do some extra testing and make sure the logging always fails gracefully.

Reply:
In the original commit, the `rescue` was after the `yield`... meaning whatever code block was passed in to be executed (aka the `yield`) may never have executed at all if the logic to set the `data` var threw. The new commit to address this comment looks insanely safe tho (assuming you actually tried it out and there are no syntax errors in the code)!

Reply:
Another way to put it, any exceptions thrown setting the `data` variable in the original commit would have caused the entire request to essentially be discarded (i think).

Reply:
yep, you're totally right. and things I tested with this latest commit:

* exception in extra_logging before yield (sets error_msg in data, then logs in `ensure`)
* exception in UsersController.user_dashboard (error propagates to `rescue` on L2443, sets error_msg in data, sets return code to 500, then logs in `ensure`)
* exception in extra_logging ensure block (when it can't run `data.to_json`, it fails gracefully and logs with a string instead)
* no exceptions (logs data in `ensure` and has 200 return code)